packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -196,104 +197,248 @@ class CupertinoAlertDialog extends StatelessWidget {
-///
-/// See also:
-///
-///  * [CupertinoAlertDialog], a dialog that informs the user about situations
-///    that require acknowledgement
-class CupertinoDialogAction extends StatelessWidget {
-  /// Creates an action for an iOS-style dialog.
-  const CupertinoDialogAction({
-    this.onPressed,
-    this.isDefaultAction = false,
-    this.isDestructiveAction = false,
-    @required this.child,
-  }) : assert(child != null);
+// iOS style layout policy widget for sizing an alert dialog's content section and
+// action button section.
+//
+// The sizing policy is partially determined by whether or not action buttons
+// are stacked vertically, or positioned horizontally. [isStacked] is used to
+// indicate whether or not the buttons should be stacked vertically.
+//
+// See [_CupertinoDialogRenderBox] for specific layout policy details.
+class _CupertinoDialogRenderWidget extends MultiChildRenderObjectWidget {
+  _CupertinoDialogRenderWidget({
+    Key key,
+    @required List<Widget> children,
+    bool isStacked = false,
+  }) : _isStacked = isStacked,
+       super(key: key, children: children);
 
-  /// The callback that is called when the button is tapped or otherwise
-  /// activated.
-  ///
-  /// If this is set to null, the button will be disabled.
-  final VoidCallback onPressed;
+  final bool _isStacked;
 
-  /// Set to true if button is the default choice in the dialog.
-  ///
-  /// Default buttons are bold.
-  final bool isDefaultAction;
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return new _CupertinoDialogRenderBox(isStacked: _isStacked);
+  }
+}
 
-  /// Whether this action destroys an object.
-  ///
-  /// For example, an action that deletes an email is destructive.
-  final bool isDestructiveAction;
+// iOS style layout policy for sizing an alert dialog's content section and action
+// button section.
+//
+// The policy is as follows:
+//
+// If all content and buttons fit on screen:
+// The content section and action button section are sized intrinsically and centered
+// vertically on screen.
+//
+// If all content and buttons do not fit on screen:
+// A minimum height for the action button section is calculated. The action
+// button section will not be rendered shorter than this minimum.  See
+// [_CupertinoDialogActionsRenderBox] for the minimum height calculation.
+//
+// With the minimum action button section calculated, the content section is
+// laid out as tall as it wants to be, up to the point that it hits the
+// minimum button height at the bottom.
+//
+// After the content section is laid out, the action button section is allowed
+// to take up any remaining space that was not consumed by the content section.
+class _CupertinoDialogRenderBox extends RenderBox
+    with ContainerRenderObjectMixin<RenderBox, MultiChildLayoutParentData>,
+    RenderBoxContainerDefaultsMixin<RenderBox, MultiChildLayoutParentData> {
+  _CupertinoDialogRenderBox({
+    List<RenderBox> children,
```

Comment:
i don't understand why this render object takes a list of children. If I understand the algorithm correctly, it only actually handles two children. Why not just have two child slots?

Reply:
I think this is my confusion about common practices with RenderBox implementations.  I looked around and seemed to see this pattern of allowing arbitrary children (probably because I was looking at generic layout RenderBoxes).

But I do have a question about this.  It looks like RenderBoxes have children set on them directly by their Widget or Element (can't remember where I saw it).  Do we need children to be taken in by the constructor at all?

Reply:
A RenderBox is in control of its child model. It should handle its children in whatever form is appropriate for what it does. If it's a render object that just renders two children, it should take two children, typically as named arguments, stored in named fields. If it can render an arbitrary number of children, it should take an arbitrary number, typically as a list, stored in a list called `children`. If it only handles one child, it should just take one, named `child`. And so on.

We have cases of zero, one, an arbitrary list, a fixed number of named children, a single child plus a list, and probably various other cases scattered throughout the codebase.

----

Regarding your second question, render objects in general don't know about widgets and need not be used with widgets. When used directly -- e.g. as in https://github.com/flutter/flutter/blob/master/examples/layers/rendering/hello_world.dart -- then it's not uncommon to provide the children in the constructors.

In your case it's a private render object just for a widget, so the constructor argument will never be used, but we still include it for consistency. It makes it easier to not mess up if we ever make it public, or if anyone copies and pastes it into public code.